### DISCUSSION TITLE
DeepWannier NN predicts 0 global dipole only, no atomic dipole

### DISCUSSION THREAD

**Discussion by alikhamze** [deepmodeling/deepmd-kit#2243] [Q&A] [answered]
Hello,

I trained a DeepWannier neural network on a small set of training data (structures generated using MD at 500 K, then Wannier center calculations done using quantum espresso) as a proof-of-concept before generating a larger training set for an actually accurate model. The training completed successfully, and I then froze and compressed the model. 
To test the model, I generated additional structures for the same system (again using MD, at 300 K this time) to apply the model to.

I performed the inference following the script from the Princeton workshop over the summer: https://github.com/CSIprinceton/workshop-july-2022/blob/main/hands-on-sessions/day-2/4-deep-wannier/dipole_distribution.py

However, when I infer from my model, I either get 0 or no output at all. When I ask for a global dipole moment only, I get 0:
![image](https://user-images.githubusercontent.com/6139089/211919476-7de1acd0-9226-454b-8619-168b2a221a44.png)

When I ask for the atomic dipoles as well, the output is blank:
![image](https://user-images.githubusercontent.com/6139089/211919643-62a8e9f8-ca5f-47cd-b04a-353f7b6c9656.png)

When I use the model distributed in the Princeton workshop with the structures provided, I do get non-zero dipole moments predicted by the model, so I don't think my DeepMD installation is the source of the problem. However, in case the information will be helpful, I am using DeepMD v2.1.5.

Here is my training input file (converted from json to txt to be uploaded to github).
[input.txt](https://github.com/deepmodeling/deepmd-kit/files/10396025/input.txt)

Does anyone have any idea why my toy model won't predict a non-zero dipole moment?

Thanks,
-Ali

**Comment by salinelake** [comment]
Hi Ali,

Can you provide a minimal example of the training set?

**Reply by alikhamze** [reply]
Hi Pinchen,

Attached is the data from the first 10 frames. Let me know if anything looks wrong?

Thanks for taking a look!


[sample_data.tar.gz](https://github.com/deepmodeling/deepmd-kit/files/10403223/sample_data.tar.gz)

**Comment by salinelake** [comment, marked as answer]
I trained a model with the data you provide. The model looks alright. You can test your model directly against the training data by 
```

from deepmd.infer import DeepDipole 
import numpy as np
import dpdata
import os

trainset = [
        "./data"
        # "/home/ali/calculations/bn/train_deepwannier/data/processed",
]
 
## load all training data
model_path = './dipole.pb'
model = DeepDipole(model_path)
## evaluate the trainset
dipole_ref = []
global_ref = []
dipole_pred = []
global_pred = []
for data_dir in trainset:
    print('loading ', data_dir) 
    cells = np.load(os.path.join(data_dir,'set.000/box.npy'))
    coords = np.load(os.path.join(data_dir,'set.000/coord.npy'))
    atypes = np.loadtxt(os.path.join(data_dir,'type.raw')).astype(int)
    dipole_ref.append(np.load(os.path.join(data_dir,'set.000/atomic_dipole.npy')).reshape(-1,3))
    global_ref.append(np.load(os.path.join(data_dir,'set.000/dipole.npy')).reshape(-1,3))
    _dipole_pred = model.eval(coords, cells, atypes)
    dipole_pred.append(_dipole_pred.reshape(-1,3))
    global_pred.append(_dipole_pred.sum(-2).reshape(-1,3))

dipole_ref = np.concatenate(dipole_ref)
dipole_pred = np.concatenate(dipole_pred)
global_ref = np.concatenate(global_ref)
global_pred = np.concatenate(global_pred)
##
print(global_ref.shape)
print(global_pred.shape)

```

I guess the fact you get either get 0 or no output following the script from the Princeton workshop is due an incorrect atypes:
```
atypes = np.array(atoms.get_atomic_numbers()) - 1
dipole = dipole_model.eval(atoms.get_positions(),atoms.get_cell(),atom_types=atypes)
```
Please check "positions" and "cell" feeded to dipole model has the correct shape. 
And make sure atom_types is given consistently with your training data.

**Reply by alikhamze** [reply]
Finally got a chance to try this and double check things, this worked! Thank you, Pinchen!
